File: java/src/main/java/problems1_100/JumpGame2.java
package problems1_100;

enum JumpGame2 {
    DEFAULT {
        @Override
        int jump(int[] nums) {
            int steps = 0;
            int maxLastJump = 0;
            int maxCurrentJump = 0;
            for (int i = 0; i < nums.length - 1; i++) {
                maxCurrentJump = Math.max(maxCurrentJump, i + nums[i]);
                if (i != maxLastJump) {
                    continue;
                }
                steps++;
                maxLastJump = maxCurrentJump;
            }
            return steps;
        }
    };

    abstract int jump(int[] nums);
}

File: kotlin/src/main/kotlin/problems1_100/JumpGame2.kt
package problems1_100

import kotlin.math.max

enum class JumpGame2 {
    DEFAULT {
        override fun jump(nums: IntArray): Int {
            var steps = 0
            var maxLastJump = 0
            var maxCurrentJump = 0
            for (i in 0 until nums.lastIndex) {
                maxCurrentJump = max(maxCurrentJump, (i + nums[i]))
                if (i != maxLastJump) {
                    continue
                }
                steps++
                maxLastJump = maxCurrentJump
            }
            return steps
        }
    },
    ;

    abstract fun jump(nums: IntArray): Int
}
